### DISCUSSION TITLE
'DT_N_S_soc_S_timers_40012c00_S_pwm' undeclared

### DISCUSSION THREAD

**Discussion by zzzh** [zephyrproject-rtos/zephyr#43743] [Q&A]
**Describe the bug**
Got compile error when trying to move code from zephyr-2.3.0 to zephyr-3.0.99

Please also mention any information which could help others to understand
the problem you're facing:
 - What target platform are you using?   ARM-STM32F0
 - What have you tried to diagnose or workaround this issue?  changing dts file
 - changed as an example in sample, but don't work

**To Reproduce**
in my project folder, run west build, then error as below would appear:

my_project_path/build/zephyr/include/generated/devicetree_unfixed.h:4751:54: error: 'DT_N_S_soc_S_timers_40012c00_S_pwm' undeclared here (not in a function); did you mean 'DT_N_S_soc_S_timers_40012c00_S_pwm_ORD'?
 4751 | #define DT_N_S_pwmleds_S_pwm_led_all_P_pwms_IDX_0_PH DT_N_S_soc_S_timers_40012c00_S_pwm

**Expected behavior**
No error

**Impact**
What impact does this issue have on your progress (e.g., annoyance, showstopper)
Can't continue

**Logs and console output**
If applicable, add console logs or other types of debug information
e.g Wireshark capture or Logic analyzer capture (upload in zip archive).
copy-and-paste text and put a code fence (\`\`\`) before and after, to help
explain the issue. (if unable to obtain text log, add a screenshot)

my_project_path/build/zephyr/include/generated/devicetree_unfixed.h:4751:54: error: 'DT_N_S_soc_S_timers_40012c00_S_pwm' undeclared here (not in a function); did you mean 'DT_N_S_soc_S_timers_40012c00_S_pwm_ORD'?
 4751 | #define DT_N_S_pwmleds_S_pwm_led_all_P_pwms_IDX_0_PH DT_N_S_soc_S_timers_40012c00_S_pwm

**Environment (please complete the following information):**
 - OS: (e.g. Linux, MacOS, Windows) Linux
 - Toolchain (e.g Zephyr SDK, ...) Zephyr SDK
 - Commit SHA or Version used   zephyr-3.0.99

**Additional context**
Add any other context that could be relevant to your issue, such as pin setting,
target configuration, ...

code in dts file:

&timers1 {
    //st,prescaler = <10000>;
    status = "okay";

    pwm1: pwm {
        status = "okay";        
        pinctrl-0 = <&tim1_ch4_pa11 &tim1_ch3n_pb15 &tim1_ch2n_pb14 &tim1_ch1n_pb13>;
        pinctrl-names = "default";
    };
};

**Comment by erwango** [comment]
Can you clarify which board you're using ? `ARM-STM32F0` is not an official board name

**Comment by zzzh** [comment]
It's not an official board, but compile well in zephyr-2.3, some pinctrl thing had been added.

**Comment by zzzh** [comment]
DT_PWMS_LABEL is deprecated, which command should be used instead?

**Comment by erwango** [comment]
> DT_PWMS_LABEL is deprecated, which command should be used instead?

It depends what you need to do. Do you have an example of usage ?

**Comment by erwango** [comment]
Should be something like `DT_PWMS_CTLR`

**Comment by zzzh** [comment]
I used DT_PWMS_CTLR to replace DT_PWMS_LABEL, seems this is the source of the error.

for example:
it was: DT_PWMS_LABEL(DT_NODALABEL(xxx)), and work well in zephyr-2.3
now is: DT_PWMS_CTLR(DT_NODALABEL(xxx)), and don't work in zephyr-3.0.99

**Comment by erwango** [comment]
I suggest to look at `samples/basic/blinky_pwm/src/main.c` to see how it is used in current code.

**Comment by zzzh** [comment]
Got it, thank you.

**Comment by zzzh** [comment]
@erwango 
I tried the way in example, it don't work.
Got undeclared issue in compile.

**Comment by erwango** [comment]
I've transfered this to discussions, as it is likely not a bug so far.
Can you provide the code you're using ?

**Reply by zzzh** [reply]
Sure, let me check it again first.

**Reply by zzzh** [reply]
Here it is:

in c file:
const struct device *dev = DEVICE_DT_GET(DT_PWMS_CTLR(DT_NODELABEL(pwm_led_red)));

in dts file:

pwmleds {
        compatible = "pwm-leds";
        pwm_led_red: led_pwm_red { 
            pwms = <&pwm1 4 0 0>;
        };
}

&timers1 {
    status = "okay";
    pwm1: pwm {
        pinctrl-0 = <&tim1_ch4_pa11 &tim1_ch3n_pb15 &tim1_ch2n_pb14 &tim1_ch1n_pb13>;
        pinctrl-names = "default";
        status = "okay";
    };
};

timers1: timers@40012c00 {
			compatible = "st,stm32-timers";
			reg = <0x40012c00 0x400>;
			clocks = <&rcc STM32_CLOCK_BUS_APB1_2 0x00000800>;
			interrupts = <13 0>, <14 0>;
			interrupt-names = "brk_up_trg_com", "cc";
			st,prescaler = <0>;
			status = "disabled";
			label = "TIMERS_1";

			pwm {
				compatible = "st,stm32-pwm";
				status = "disabled";
				label = "PWM_1";
				#pwm-cells = <3>;
			};
		};

compile error:
'DT_N_S_soc_S_timers_40012c00_S_pwm' undeclared

**Reply by erwango** [reply]
Can you try to compile `samples/basic/fade_led` using board nucleo_f031k6?
This is using the similar code on same target. This is compiling on my side.
- If not working you likely have some bad addition in your code.
- if working, have a try using your own board and let me know

**Reply by zzzh** [reply]
I compiled samples/basic/fade_led using board nucleo_f031k6, it works.
through, seems the same way don't work on my board.

**Reply by zzzh** [reply]
Got same error as before

**Reply by erwango** [reply]
ok, then I suggest comparing `nucleo_f031k6` with your board. Unfortunately, I won't be able to help here.

**Reply by erwango** [reply]
@zzzh You mention you have the following in your dts file:
```
timers1: timers@40012c00 {
compatible = "st,stm32-timers";
reg = <0x40012c00 0x400>;
clocks = <&rcc STM32_CLOCK_BUS_APB1_2 0x00000800>;
interrupts = <13 0>, <14 0>;
interrupt-names = "brk_up_trg_com", "cc";
st,prescaler = <0>;
status = "disabled";
label = "TIMERS_1";
```
Is that in your board dts file ? 
Normally, you shouldn't need to add it as it should be provided in soc.dtsi file.

**Reply by zzzh** [reply]
@erwango It's from the dtsi file, just put it there to make things clear.

**Reply by erwango** [reply]
ok. Don't hesitate to provide maximum details in problem description.
The kind of issue you're facing is one of those hiding in such details.
